working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,923 @@
+This answer is used to select appropriate control flow in the surrounding
+construct. For example, in some hypothetical match statement, if the pattern
+does not match, then the case containing the pattern is skipped.
+
+```
+matcher ::=
+  | wildcardMatcher
+  | literalMatcher
+  | constantMatcher
+  | listMatcher
+  | mapMatcher
+  | typedMatcher
+  | bindMatcher
+  | typeTestMatcher
+  | tupleMatcher
+  | extractMatcher
+
+matchers ::= matcher ( "," matcher )* ","?
+```
+
+**TODO: Do set matchers make sense?**
+
+### Wildcard matcher
+
+```
+wildcardMatcher ::= "_"
+```
+
+A wildcard pattern always matches. This is useful in places where a subpattern
+is required but you always want to succeed. It can function as a "default"
+pattern for the last case in a pattern matching statement.
+
+### Literal matcher
+
+```
+literalMatcher ::=
+  | booleanLiteral
+  | numericLiteral
+  | stringLiteral
+```
+
+A literal pattern determines if the value is equivalent to the given literal
+value. The pattern matches if the two values are equal, determined by passing
+the pattern's literal value to the `==` method on the value being matched.
+
+### Constant matcher
+
+```
+constantMatcher ::= identifier ( "." identifier ( "." identifier )? )?
+```
+
+Determines if the value is equivalent to the value of the given constant
+expression. The expression is syntactically restricted to be either:
+
+*   **A bare identifier.** In this case, the identifier must resolve to a
+    constant declaration in scope.
+
+*   **A prefixed or qualified identifier.** In other words, `a.b`. It must
+    resolve to either a top level constant imported from a library with a
+    prefix, a static constant in a class, or an enum case.
+
+*   **A prefixed qualified identifier.** Like `a.B.c`. It must resolve to an
+    enum case on an emum type that was imported with a prefix.
+
+The const expression is evaluated and the result passed to the matched value's
+`==` operator to determine if it matches.
+
+**TODO: Do we want to allow other kinds of constant expressions like `1 + 2`?
+Switch statements allow arbitrary const expressions. We probably don't want to
+allow const collection literals since those clash with the corresponding
+collection patterns.**
+
+### List matcher
+
+```
+listMatcher ::= "[" matchers "]"
+```
+
+Matches and destructures lists. If the value's type does not implement `List<T>`
+for some T, then the match fails. If the length of the list determined by
+calling `length` is not equal to the number of subpatterns, the match fails.
+
+Otherwise, extracts a series of elements from the list using the index operator
+and tests them against the corresponding subpatterns. The list matcher succeeds
+if all of the subpatterns match.
+
+**TODO: Allow a `...` element to skip over elements and match suffixes. Can you
+capture the rest in a variable?**
+
+**TODO: Should we loosen this to match against any Iterables by using
+`elementAt()`?**
```

Comment:
One nice thing about generalizing this to Iterables is that it would let you match sets using this too. My feeling is that allowing arbitrary Iterables to be matched is more good than bad. When a user uses a `[ ... ]` pattern, they probably don't care that it's *literally* a List (which could still be any random wild user-defined class implementing List!), they just want to destructure an indexable thing.

Reply:
I'm a little on the fence here, and I think it comes down to a bit of asymmetry around who controls the matching: the thing being matched, or the pattern.  This is reflected in my comments around which thing is the receiver of the `==` method.  Which one you choose changes who is in control of the match.  Similarly, I think the question here is, does a list pattern ask the question "Are you a list (or perhaps iterable)" vs "do you match a list pattern".  The latter question could be answered yes by anything implementing a more limited interface exposing only length and indexing, or structurally by checking for an "asList" method, or it could be simply delegated to the scrutinee by calling a `match` method which takes a pattern object as an argument and decides what to do about it (i.e. either to agree to be matched by that pattern, filling in the appropriate sub-bits, or decline).  If all the user cares about is destructuring an indexable thing, then why require it even to be an iterable?

Pragmatically, one key difference here is whether you get to promote the scrutinee to a specific match type (or alternatively, bind the scrutinee to a new value with a refined type).